File: app/stats.py
"""Reading statistics store (statistics.json).

Schema:
  {"version": 1,
   "books": {"<book_id>": {"total_seconds": int, "sessions": int,
                           "pages_flipped": int, "last_read_at": "ISO",
                           "days": {"YYYY-MM-DD": {"seconds": int, "pages": int}}}},
   "global": {"total_seconds": int, "days": {...}}}
"""
import json
import threading
from datetime import date, timedelta
from pathlib import Path
from typing import Optional

from .storage import atomic_write_json, now_iso


def _local_today() -> str:
    return date.today().isoformat()


def _bump_day(days: dict, day: str, seconds: int, pages: int) -> None:
    entry = days.setdefault(day, {"seconds": 0, "pages": 0})
    entry["seconds"] += seconds
    entry["pages"] += pages


def _streak_days(days: dict) -> int:
    d = date.today()
    if days.get(d.isoformat()) is None:
        d -= timedelta(days=1)
    streak = 0
    while True:
        entry = days.get(d.isoformat())
        if not entry or int(entry.get("seconds", 0)) <= 0:
            break
        streak += 1
        d -= timedelta(days=1)
    return streak


def _enrich(rec: dict) -> dict:
    total = int(rec.get("total_seconds", 0))
    sessions = int(rec.get("sessions", 0))
    days = rec.get("days") or {}
    today = _local_today()
    week_start = (date.today() - timedelta(days=6)).isoformat()
    today_secs = int(days.get(today, {}).get("seconds", 0))
    today_pages = int(days.get(today, {}).get("pages", 0))
    week_secs = sum(
        int(v.get("seconds", 0))
        for k, v in days.items()
        if k >= week_start and k <= today
    )
    return {
        **rec,
        "days": dict(days),
        "total_seconds": total,
        "sessions": sessions,
        "pages_flipped": int(rec.get("pages_flipped", 0)),
        "last_read_at": rec.get("last_read_at", ""),
        "today_seconds": today_secs,
        "today_pages": today_pages,
        "week_seconds": week_secs,
        "avg_session_seconds": round(total / sessions) if sessions else 0,
        "streak_days": _streak_days(days),
    }


class StatsStore:
    def __init__(self, file: Path):
        self._file = file
        self._lock = threading.RLock()
        self._books: dict = {}
        self._global: dict = {}

    def load(self) -> None:
        try:
            with open(self._file, encoding="utf-8") as f:
                data = json.load(f)
            self._books = data.get("books", {})
            self._global = data.get("global", {})
        except (OSError, json.JSONDecodeError, AttributeError):
            self._books = {}
            self._global = {}

    def save(self) -> None:
        with self._lock:
            atomic_write_json(
                self._file,
                {"version": 1, "books": self._books, "global": self._global},
            )

    def record_reading(self, book_id: str, seconds: int, pages_flipped: int = 0) -> None:
        seconds = max(0, int(seconds))
        pages = max(0, int(pages_flipped))
        today = _local_today()
        with self._lock:
            b = self._books.setdefault(
                book_id,
                {
                    "total_seconds": 0,
                    "sessions": 0,
                    "pages_flipped": 0,
                    "last_read_at": "",
                    "days": {},
                },
            )
            b.setdefault("days", {})
            b["total_seconds"] += seconds
            if pages:
                b["pages_flipped"] += pages
            if seconds > 0:
                b["sessions"] += 1
                b["last_read_at"] = now_iso()
                _bump_day(b["days"], today, seconds, pages)
            g = self._global
            g.setdefault("days", {})
            g["total_seconds"] = g.get("total_seconds", 0) + seconds
            if seconds > 0:
                _bump_day(g["days"], today, seconds, pages)
            self.save()

    def get_book(self, book_id: str) -> dict:
        with self._lock:
            b = self._books.get(book_id)
            if not b:
                return {
                    "total_seconds": 0,
                    "sessions": 0,
                    "pages_flipped": 0,
                    "last_read_at": "",
                    "days": {},
                    "today_seconds": 0,
                    "today_pages": 0,
                    "week_seconds": 0,
                    "avg_session_seconds": 0,
                    "streak_days": 0,
                }
            return _enrich(b)

    def get_global(self) -> dict:
        with self._lock:
            return _enrich(self._global)

    def remove_book(self, book_id: str) -> None:
        with self._lock:
            if book_id in self._books:
                del self._books[book_id]
                self.save()

File: android/app/src/main/java/io/github/gighi947/ankeshelf/data/Stats.kt
package io.github.gighi947.ankeshelf.data

import kotlinx.serialization.Serializable
import java.io.File
import java.time.LocalDate
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

@Serializable
data class DayEntry(
    val seconds: Int = 0,
    val pages: Int = 0,
)

@Serializable
data class StatsEntry(
    val total_seconds: Int = 0,
    val sessions: Int = 0,
    val pages_flipped: Int = 0,
    val last_read_at: String = "",
    val days: Map<String, DayEntry> = emptyMap(),
)

@Serializable
data class StatsFile(
    val version: Int = 1,
    val books: Map<String, StatsEntry> = emptyMap(),
    val global: StatsEntry = StatsEntry(),
)

data class EnrichedStats(
    val total_seconds: Int = 0,
    val sessions: Int = 0,
    val pages_flipped: Int = 0,
    val last_read_at: String = "",
    val days: Map<String, DayEntry> = emptyMap(),
    val today_seconds: Int = 0,
    val today_pages: Int = 0,
    val week_seconds: Int = 0,
    val avg_session_seconds: Int = 0,
    val streak_days: Int = 0,
)

/** 阅读统计（statistics.json）：全局 + 每书 + 按天。 */
class StatsStore(private val file: File) {

    private val lock = ReentrantLock()
    private var books: MutableMap<String, StatsEntry> = mutableMapOf()
    private var global: StatsEntry = StatsEntry()

    fun load() {
        val data = when (val r = readJsonStore<StatsFile>(file)) {
            is StoreLoadResult.Ok -> r.value
            StoreLoadResult.Missing -> StatsFile()
            is StoreLoadResult.Corrupt -> {
                logWarn("AnkeShelf", "statistics.json 损坏，回退默认：${r.detail}")
                StatsFile()
            }
            is StoreLoadResult.IoError -> {
                logWarn("AnkeShelf", "statistics.json 读取失败：${r.detail}")
                StatsFile()
            }
        }
        lock.withLock {
            books = data.books.toMutableMap()
            global = data.global
        }
    }

    fun save() {
        val snapshot = lock.withLock { books.toMap() to global }
        atomicWriteJson(
            file,
            Shelf.json.encodeToString(
                StatsFile.serializer(),
                StatsFile(books = snapshot.first, global = snapshot.second),
            ),
        )
    }

    fun recordReading(bookId: String, seconds: Int, pagesFlipped: Int = 0) {
        val secs = maxOf(0, seconds)
        val pages = maxOf(0, pagesFlipped)
        val today = LocalDate.now().toString()
        lock.withLock {
            var b = books[bookId] ?: StatsEntry()
            var day = b.days[today] ?: DayEntry()
            b = b.copy(
                total_seconds = b.total_seconds + secs,
                pages_flipped = if (pages > 0) b.pages_flipped + pages else b.pages_flipped,
                sessions = if (secs > 0) b.sessions + 1 else b.sessions,
                last_read_at = if (secs > 0) nowIso() else b.last_read_at,
            )
            if (secs > 0) {
                day = DayEntry(seconds = day.seconds + secs, pages = day.pages + pages)
                b = b.copy(days = b.days + (today to day))
            }
            books[bookId] = b

            var g = global
            var gDay = g.days[today] ?: DayEntry()
            g = g.copy(
                total_seconds = g.total_seconds + secs,
                pages_flipped = if (pages > 0) g.pages_flipped + pages else g.pages_flipped,
                sessions = if (secs > 0) g.sessions + 1 else g.sessions,
                last_read_at = if (secs > 0) nowIso() else g.last_read_at,
            )
            if (secs > 0) {
                gDay = DayEntry(seconds = gDay.seconds + secs, pages = gDay.pages + pages)
                g = g.copy(days = g.days + (today to gDay))
            }
            global = g
            save()
        }
    }

    fun getBook(bookId: String): EnrichedStats =
        lock.withLock { enrich(books[bookId] ?: StatsEntry()) }

    fun getGlobal(): EnrichedStats = lock.withLock { enrich(global) }

    fun removeBook(bookId: String) {
        lock.withLock {
            if (books.remove(bookId) != null) save()
        }
    }

    private fun enrich(rec: StatsEntry): EnrichedStats {
        val today = LocalDate.now().toString()
        val weekStart = LocalDate.now().minusDays(6).toString()
        val todaySecs = rec.days[today]?.seconds ?: 0
        val todayPages = rec.days[today]?.pages ?: 0
        val weekSecs = rec.days
            .filterKeys { it >= weekStart && it <= today }
            .values.sumOf { it.seconds }
        return EnrichedStats(
            total_seconds = rec.total_seconds,
            sessions = rec.sessions,
            pages_flipped = rec.pages_flipped,
            last_read_at = rec.last_read_at,
            days = rec.days,
            today_seconds = todaySecs,
            today_pages = todayPages,
            week_seconds = weekSecs,
            avg_session_seconds = if (rec.sessions > 0) kotlin.math.round(rec.total_seconds.toDouble() / rec.sessions).toInt() else 0,
            streak_days = streakDays(rec.days),
        )
    }

    private fun streakDays(days: Map<String, DayEntry>): Int {
        var d = LocalDate.now()
        if (days[d.toString()] == null) d = d.minusDays(1)
        var streak = 0
        while (true) {
            val entry = days[d.toString()] ?: break
            if (entry.seconds <= 0) break
            streak++
            d = d.minusDays(1)
        }
        return streak
    }
}
